sw/device/silicon_creator/lib/otbn_util.c

```diff
@@ -0,0 +1,125 @@
+  };
+}
+
+uint32_t otbn_get_internal_err_bits(otbn_t *ctx) { return ctx->error_bits; }
+
+rom_error_t otbn_func_ptr_to_imem_addr(const otbn_t *ctx, otbn_ptr_t ptr,
+                                       uint32_t *imem_addr_otbn) {
+  const uint32_t *app_imem_start_addr = ctx->app.imem_start;
+  const uint32_t *app_imem_end_addr = ctx->app.imem_end;
+
+  if (ptr < app_imem_start_addr || ptr > app_imem_end_addr) {
+    return kErrorOtbnInvalidArgument;
+  }
+  *imem_addr_otbn = (ptrdiff_t)ptr - (ptrdiff_t)app_imem_start_addr;
+  return kErrorOk;
+}
+
+rom_error_t otbn_data_ptr_to_dmem_addr(const otbn_t *ctx, otbn_ptr_t ptr,
+                                       uint32_t *dmem_addr_otbn) {
+  const uint32_t *app_dmem_start_addr = ctx->app.dmem_start;
+  const uint32_t *app_dmem_end_addr = ctx->app.dmem_end;
+
+  if (ptr < app_dmem_start_addr || ptr > app_dmem_end_addr) {
+    return kErrorOtbnInvalidArgument;
+  }
+  *dmem_addr_otbn = (ptrdiff_t)ptr - (ptrdiff_t)app_dmem_start_addr;
+  return kErrorOk;
+}
+
+rom_error_t otbn_busy_wait_for_done(otbn_t *ctx) {
+  while (otbn_is_busy()) {
+  }
+
+  otbn_err_bits_t err_bits;
+  otbn_get_err_bits(&err_bits);
+  if (err_bits != kOtbnErrBitsNoError) {
+    ctx->error_bits = err_bits;
+    return kErrorOtbnExecutionFailed;
+  }
+  return kErrorOk;
+}
+
+rom_error_t otbn_load_app(otbn_t *ctx, const otbn_app_t app) {
+  if (app.imem_end <= app.imem_start || app.dmem_end < app.dmem_start) {
+    return kErrorOtbnInvalidArgument;
+  }
+
+  const size_t imem_size = app.imem_end - app.imem_start;
+  const size_t dmem_size = app.dmem_end - app.dmem_start;
+
+  RETURN_IF_ERROR(otbn_imem_write(0, app.imem_start, imem_size));
+
+  if (dmem_size > 0) {
+    RETURN_IF_ERROR(otbn_dmem_write(0, app.dmem_start, dmem_size));
+  }
+
+  ctx->app = app;
+  ctx->app_is_loaded = true;
+  return kErrorOk;
+}
+
+rom_error_t otbn_call_function(otbn_t *ctx, otbn_ptr_t func) {
+  if (!ctx->app_is_loaded) {
+    return kErrorOtbnInvalidArgument;
+  }
+
+  uint32_t func_imem_addr;
+  RETURN_IF_ERROR(otbn_func_ptr_to_imem_addr(ctx, func, &func_imem_addr));
+  RETURN_IF_ERROR(otbn_start(func_imem_addr));
+
+  return kErrorOk;
+}
+
+rom_error_t otbn_copy_data_to_otbn(otbn_t *ctx, size_t len, const void *src,
+                                   otbn_ptr_t dest) {
+  uint32_t dest_dmem_addr;
+  RETURN_IF_ERROR(otbn_data_ptr_to_dmem_addr(ctx, dest, &dest_dmem_addr));
+  RETURN_IF_ERROR(otbn_dmem_write(dest_dmem_addr, src, len));
+
+  return kErrorOk;
+}
+
+rom_error_t otbn_copy_data_from_otbn(otbn_t *ctx, size_t len_bytes,
+                                     otbn_ptr_t src, void *dest) {
+  uint32_t src_dmem_addr;
+  RETURN_IF_ERROR(otbn_data_ptr_to_dmem_addr(ctx, src, &src_dmem_addr));
+  RETURN_IF_ERROR(otbn_dmem_read(src_dmem_addr, dest, len_bytes));
+
+  return kErrorOk;
+}
+
+rom_error_t otbn_zero_data_memory(otbn_t *ctx) {
+  const size_t dmem_size_words = kOtbnDMemSizeBytes / sizeof(uint32_t);
+  bool retval = kErrorOk;
```

Comment:
`bool` seems like a type error here. It also seems like it's never written to...?